src/collections/componentSet.ts

```diff
@@ -43,6 +43,78 @@ export type RetrieveSetOptions = Omit<MetadataApiRetrieveOptions, 'components'>;
 
 const KEY_DELIMITER = '#';
 
+/**
+ * This is an extension of the Map class that treats keys as the same by matching first normally,
+ * then decoded. Decoding the key before comparing can solve some edge cases in component fullNames
+ * such as Layouts. See: https://github.com/forcedotcom/cli/issues/1683
+ *
+ * Examples:
+ *
+ * Given a map with entries:
+ * ```javascript
+ * 'layout#Layout__Broker__c-v1%2E1 Broker Layout' : {...}
+ * 'layout#Layout__Broker__c-v9.2 Broker Layout' : {...}
+ * ```
+ *
+ * `decodeableMap.has('layout#Layout__Broker__c-v1.1 Broker Layout')` --> returns `true`
+ * `decodeableMap.has('layout#Layout__Broker__c-v9%2E2 Broker Layout')` --> returns `true`
+ */
+class DecodeableMap<K extends string, V> extends Map<string, V> {
```

Comment:
suggestion: put this in a separate file and import into CS.  CS is a big file already.

then it's also be easier to throw some UT at without the complexity of CS

Reply:
I like it.  Moving.